src/Conversion/ONNXToKrnl/Math/LRN.cpp

```diff
@@ -55,87 +55,86 @@ struct ONNXLRNOpLowering : public ConversionPattern {
-    Value squareVal = create.math.mul(loadVal, loadVal);
-
-    Value sumValue = create.krnl.load(sumAlloc, ArrayRef<Value>{});
-    sumValue = create.math.add(sumValue, squareVal);
-    create.krnl.store(sumValue, sumAlloc, ArrayRef<Value>{});
-
-    // Compute and store the output
-    // y = x / ((bias + (alpha / nsize) * square_sum) ** beta)
-    rewriter.restoreInsertionPoint(outputLoopBody);
-    SmallVector<Value, 4> storeIndices;
-    for (int i = 0; i < outputRank; ++i) {
-      storeIndices.emplace_back(outputLoops.getInductionVar(i));
-    }
-    Value xValue = create.krnl.load(input, storeIndices);
-    sumValue = create.krnl.load(sumAlloc);
-    Value tempValue =
-        create.math.pow(create.math.add(biasValue,
-                            create.math.mul(alphaDivSizeValue, sumValue)),
-            betaValue);
-    Value resultValue = create.math.div(xValue, tempValue);
-
-    create.krnl.store(resultValue, alloc, storeIndices);
+    KrnlBuilder createKrnl(rewriter, loc);
+    ValueRange outputLoopDef = createKrnl.defineLoops(outputRank);
+    SmallVector<IndexExpr, 4> lbs(outputRank, LiteralIndexExpr(0));
+    createKrnl.iterateIE(outputLoopDef, outputLoopDef, lbs,
+        shapeHelper.dimsForOutput(0),
+        [&](KrnlBuilder &createKrnl, ValueRange outputLoopInd) {
+          // Insert computation of square_sum.
+          // square_sum[n, c, d1, ..., dk] = sum(X[n, i, d1, ..., dk] ^ 2),
+          // where max(0, c - floor((size - 1) / 2)) <= i
+          // and i<= min(C - 1, c + ceil((size - 1) / 2)).
+
+          // Get a child IndexExpr context.
+          IndexExprScope childScope(&rewriter, shapeHelper.scope);
+
+          // Compute the lower bound and upper bound for square_sum.
+          const int loopIndexForC = 1;
+          DimIndexExpr cIE(outputLoopInd[loopIndexForC]);
+          MemRefBoundsIndexCapture inputBounds(input);
+          DimIndexExpr CIE(inputBounds.getDim(loopIndexForC));
+          SymbolIndexExpr sizeIE = LiteralIndexExpr(sizeLit);
+
+          SmallVector<IndexExpr, 2> lbMaxList;
+          lbMaxList.emplace_back(LiteralIndexExpr(0));
+          lbMaxList.emplace_back(
+              cIE - (sizeIE - 1).floorDiv(LiteralIndexExpr(2)));
+
+          SmallVector<IndexExpr, 2> ubMinList;
+          ubMinList.emplace_back(CIE);
+          ubMinList.emplace_back(
+              cIE + 1 + (sizeIE - 1).ceilDiv(LiteralIndexExpr(2)));
+
+          // Initialize sum, single scalar, no need for default alignment.
+          MemRefType scalarMemRefType = MemRefType::get({}, elementType, {}, 0);
+          MultiDialectBuilder<KrnlBuilder, MemRefBuilder, MathBuilder> create(
+              rewriter, loc);
+
+          Value sumAlloc = create.mem.alloc(scalarMemRefType);
+          createKrnl.store(create.math.constant(elementType, 0), sumAlloc);
+
+          // Create the sum reduction loop
+          ValueRange sumLoopDef = createKrnl.defineLoops(1);
+          Value sumValue;
```

Comment:
I think you `sumValue` should be declared in the loop at line 117 given that every loop iteration initializes it and is not live on exit.

Reply:
It is live out, it is used at line 132.

Reply:
Line 129 redefines it before it is used at line 132:
```
129  sumValue = createKrnl.load(sumAlloc);
```